### ISSUE TITLE
send spaces instead of tabs to the front-end

### ISSUE THREAD

**Issue by eyelash** [google/xi-editor#467]
It would be nice if xi-core could send spaces to the front-end instead of tabs, so the front-end doesn't have to deal with tabs and their width. This would simplify the front-ends and the core needs to know about the tab width anyway to make moving the cursor up and down on lines with different indentation work (which doesn't work right now).

**Comment by cmyr** [comment]
@eyelash doesn't it do this already? Unless I'm misunderstanding; but there's a config setting `translate_tabs_to_spaces` which is on by default.

**Comment by raphlinus** [comment]
This is for when the file has tabs in it. There's been some discussion (I'm not sure if it's in any issue) but we haven't made a firm decision what to do yet. I'm definitely sympathetic to the problem being solved.

One downside to sending spaces is that tabs may have a different visual representation (eg when selected). But if we do translation in core at render time we can just send a style for that.

**Comment by euclio** [comment]
This is a problem for terminal frontends in general, which would need to always display tabs as spaces to ensure the correct tab width.

**Comment by jansol** [comment]
It also falls apart when using proportional fonts, where tabstops AFAIK usually have a width measured in pt, not characters. (At least that seems to be the case in the fancier word processors I've used)

**Comment by eyelash** [comment]
@jansol good point. The core still needs to know about tabs, though, since it can't measure tabs using `measure_widths` because the width of a tab depends on its surroundings. So essentially the core needs to send tabs to the front-end but it still needs to know how exactly tabs are displayed by the front-end in order to measure everything correctly.

**Comment by cmyr** [comment]
@eyelash also a good point, and something I don't think we currently handle, and which I'm not sure can be handled correctly by the current `measure_width` RPC, which assumes a given glyph run will always have the same rendered width for a given set of font/style information.

**Comment by cmyr** [comment]
(cc @raphlinus)
